### ISSUE TITLE
[docker] initial superuser created every time container starts

### ISSUE THREAD

**Issue by kzshantonu** [tubearchivist/tubearchivist#261]
So it appears that the superuser is created on container startup every time based on variables `TA_USERNAME` and `TA_PASSWORD`, even if:
- the container is started for the non-first time
- even if another superuser exists
- even if the initial user is manually deleted using the admin interface

One would presume from the wording on README, i.e. `Change the environment variables TA_USERNAME and TA_PASSWORD to create the initial credentials` and the example compose file, i.e. `your initial TA credentials` that it's only created when the container is run for the first time and/or if no other superuser exists. I believe this should be more clear in the README, and/or make the app not create a superuser if it's not the first time being run

**Comment by bbilly1** [comment]
also related: https://github.com/tubearchivist/tubearchivist/issues/82

**Comment by lamusmaser** [comment]
Seems like reasonable responses, but that is ultimately a design decision. @bbilly1 - if you agree with these, we can start to work on code and documentation updates to reflect the three use cases provided. I'm interested if you disagree with any of the points.

**Comment by dshoreman** [comment]
Found this issue while searching something unrelated, but I second @kzshantonu's expectations above. In my case, I did the first run with the vars set to `tubearchivist`/`changeme`, then changed both in the admin panel after.

I tried removing the `TA_USERNAME` and `TA_PASSWORD` vars from my compose a few days ago thinking it was their very presence causing the account to be recreated, but TA refused to start without them.

With that in mind I'd personally add one more, but (if it's even possible) it might be better as a separate issue:

4. If the container is started and there's already a superuser, you'd expect `TA_USERNAME` and `TA_PASSWORD` to be optional

**Comment by lamusmaser** [comment]
Because the `TA_USERNAME` and `TA_PASSWORD` inputs are processed by `run.sh`, before the application is aware of hte environment variables being present, I am not sure how well your recommendation would work, @dshoreman. I do 
see the utility in it, but it would take a fair amount more rearchitecting.
https://github.com/tubearchivist/tubearchivist/blob/b819c2f72371e7bc528fb97555719d0602c90b72/docker_assets/run.sh#L8-L14

From a reasonability aspect, though, I would state that this is unnecessary. Based on the model of how containers work, you should always provide the necessary elements for starting the container; modifications to deviate away from this will either require that `TA_USERNAME` and `TA_PASSWORD` are removed entirely, which means that instantiation will always have a default username and password, or the container retains these values and we have the other cases that have to be managed.

**Comment by dshoreman** [comment]
I wasn't sure either. The codebase is still mostly unknown to me, so the suggestion was more as a possible alternative/workaround in case the main issue couldn't be fixed for any reason. *That said...*

Reading through the `run.sh` after some sleep, it looks like **the superuser is created in [the same script][1]** which makes it *a lot* easier than I initially thought. *It might not need to know* if there *is* a superuser (for the most part) because *it can skip the command entirely* when some condition is met.

I need to setup a test env first but I'll put a PR together once the idea isn't just in my head. It wouldn't solve *all* the issues (e.g. auto creating if you delete all the superusers) but it *would* solve the recreation issue.

[1]: https://github.com/tubearchivist/tubearchivist/blob/b819c2f72371e7bc528fb97555719d0602c90b72/docker_assets/run.sh#L43

**Comment by dshoreman** [comment]
Think #263 should solve this now; @kzshantonu are you able to do a test build to verify?

Points 2 and 3 above are true in the PR, but point 1 can't be achieved automatically without some way of counting/listing superusers via manage.py. That way may or may not already exist, I don't know.

I backtracked on making `TA_PASSWORD` optional. Removing it causes heaps of recursion (presumably the re-architecting mentioned earlier), and it's needed for other things too:
https://github.com/tubearchivist/tubearchivist/blob/f7f4ca7728f421758179df51a64c18c6b9175bfb/tubearchivist/config/settings.py#L27-L28

**Comment by bbilly1** [comment]
Thanks everybody for looking into this. Just to clarify, the user doesn't get recreated if it already exists, the command runs at startup, but won't do anything if the user is already there. For example:
```bash
export DJANGO_SUPERUSER_PASSWORD=123456 && python manage.py createsuperuser --noinput --name "$TA_USERNAME"
```
when the username already exists will return 
```
CommandError: Error: That name is already taken.
```
and the run.sh script will just continue.

But with the rest, we are on the same page. I think the file lock approach is not ideal, for example if somebody deletes the only superuser from the admin interface, you'll have to know that there is a lock file you need to delete before you restart. 

Also this will require the hardcoded */cache* path there, before the application config can build that path. I know this is a Docker application and there are no other supported installations, but I'd rather avoid that if possible, even just from an application architecture point of view. 

I think the cleanest approach would be to create a management.py command like `python manage.py superuserexists` that returns true/false where the logic then can call the `createsuperuser` command if needed. 

This would cover all cases, and would be a fix for all. With Django, that is quite straight forward to do. 

What do you think? @dshoreman Do you want to look into it? Otherwise I can push a commit to your PR later this week implementing that command.

**Comment by kzshantonu** [comment]
@dshoreman I'd love to test. Testing branch?

**Comment by dshoreman** [comment]
@bbilly1 thanks for all your work creating/maintaining this project.

To solve the hardcoded */cache* issue, I've pushed a commit that'll fall back to the *current dir* when */cache* doesn't exist.  
Given it's unsupported anyway, will that be enough? We could even fall back to the script path, but if they're running outside Docker they might not even be *using* `./run.sh`, so we probably don't need to worry too much.

The second `celery` command also uses */cache* as a hard-coded default, should I fix that in the PR while we're at it?
https://github.com/tubearchivist/tubearchivist/blob/b819c2f72371e7bc528fb97555719d0602c90b72/docker_assets/run.sh#L48-L49


I think the other concerns actually stem from a different problem. To elaborate...

1. **What if the last superuser gets removed?**
    This sounds like alarm bells; **deleting the only superuser *shouldn't be possible*** from the admin panel. If it *is* possible, I can create a separate bug for it but I don't think that issue should stand in the way of progress here.

    If we pretend for a moment that self-annihilation is *not* possible:
    * The first start will be the *only* time there are no superusers
       * ...unless initial creation fails, but it'll run again next time until it succeeds
    * **"Expectation 1" from [earlier comments][1] is no longer a concern**
    * Checking some "superuserexists" command becomes redundant
    * If someone's daft enough to delete manually from DB, an FAQ update should suffice

2. **Users need to know about a file**
    * Lockfile path is already logged on every non-first start, so users *will* know
    * If we can fix the accidental-deletion issue, there's no other reason to ever force re-creation
    * If someone *does* still want to force it for any reason, the log output tells them how

---

With the path issue (potentially) solved, if I can look into fixing the deletion problem, would that make the lockfile approach any more appealing? If it solves the problem then (apart from being done already) it's far less code to maintain and less can go wrong. An extra command would only serve a single purpose, and it's entirely unnecessary once the deletion issue is solved.

I haven't checked yet, but if you still don't like the standard lockfile approach then we could be able to fix the issue directly instead. A new flag like `manage.py createsuperuser --initial [...]` could tell it to simply return early if there's already a superuser. I still think the lockfile is a cleaner approach, but adding a flag to the existing command would be a good middleground if you prefer.

Still suffering a bit with a cough/cold/migraine (spent all day yesterday trying to write this and failed), but let me know your thoughts and I'll see what I can do :)

[1]: https://github.com/tubearchivist/tubearchivist/issues/261#issuecomment-1155515921

**Comment by kzshantonu** [comment]
@dshoreman perfect. Behaves as expected now. Warning in log is very clear, too

**Comment by bbilly1** [comment]
@dshoreman Your reasoning is sound, let's do this with your approach with the lock file. And yes, for the celery path, I tried to fallback to a env var there, undocumented but configurable.

I know, it's a bit silly... But if you want to make the change there too in this PR, that would be excellent. So this `celerybeat-schedule` will get created at the same place as the lock file if there isn't a /cache folder. 

Then for deleting the only superuser, maybe if somebody wants to do that, what can we do. They can delete the lockfile and restart, and they are good to go again. Maybe the rest is just overthinking it. 

Thanks a lot for your help! 

Hit me up on Discord, there are lot's of more opportunities to shine in this project and I'm always looking for more people to help.

**Comment by bbilly1** [comment]
Yes, I've wiped it, it created the first user without issue, now on every restart the message shows, so I think that is working as intended. 

Thank you very much for your hard work. Please reach out on Discord, I have some more ideas. :-)

**Comment by bbilly1** [comment]
Thanks to everybody helping with that. This is now merged in v0.1.7, please update.
